router/src/server/http/write/single_tenant/auth.rs

```diff
@@ -0,0 +1,345 @@
+//! Authorization of HTTP requests using the authz service client.
+
+use std::sync::Arc;
+
+use authz::{
+    self, http::AuthorizationHeaderExtension, Action, Authorizer, Error, Permission, Resource,
+};
+use base64::{prelude::BASE64_STANDARD, Engine};
+use data_types::NamespaceName;
+use hyper::{header::HeaderValue, Body, Request};
+
+fn extract_header_token(header_value: &'_ HeaderValue) -> Option<Vec<u8>> {
+    let mut parts = header_value.as_bytes().splitn(2, |&v| v == b' ');
+    match parts.next()? {
+        b"Token" | b"Bearer" => parts
+            .next()
+            .filter(|b| !b.is_empty())
+            .map(|bytes| bytes.to_vec()),
+        b"Basic" => {
+            let value = parts.next()?;
+            if let Ok(value) = BASE64_STANDARD.decode(value) {
+                value
+                    .splitn(2, |&v| v == b':')
+                    .nth(1)
+                    .filter(|b| !b.is_empty())
+                    .map(|bytes| bytes.to_vec())
+            } else {
+                None
+            }
+        }
+        _ => None,
+    }
```

Comment:
You can reduce the duplication and simplify this a little:

```suggestion
    let token = match parts.next()? {
        b"Token" | b"Bearer" => parts.next()?.to_vec(),
        b"Basic" => parts
            .next()
            .map(|v| BASE64_STANDARD.decode(v).ok())
            .flatten()?
            .splitn(2, |&v| v == b':')
            .nth(1)?
            .to_vec(),
        _ => return None,
    };

    if token.is_empty() {
        return None;
    }
    Some(token)
```

Check the docs for the various combinators - learning them really helps write concise, readable code and it almost always outperforms alternative implementations :+1: